ssr/src/component/claim_tokens.rs

```diff
@@ -1,24 +1,89 @@
 use candid::Principal;
 use leptos::*;
+use leptos_router::*;
 
 use crate::{
-    state::canisters::authenticated_canisters,
-    utils::route::failure_redirect,
+    page::wallet::tokens::token_metadata_or_fallback,
+    state::canisters::{authenticated_canisters, unauth_canisters},
     utils::{time::sleep, token::claim_tokens_from_first_neuron_if_required},
 };
 use web_time::Duration;
 
 #[component]
-pub fn ClaimTokensOrRedirectError(token_root: Principal) -> impl IntoView {
+fn TokenClaimFallback() -> impl IntoView {
+    view! {
+        <p>"Loading..."</p>
+    }
+}
+
+#[component]
+fn TokenClaiming() -> impl IntoView {
+    view! {
+        <p>"Claiming"</p>
+    }
+}
+
+#[component]
+fn TokenDisplay(user_principal: Principal, token_root: Principal) -> impl IntoView {
+    let cans = unauth_canisters();
+    let info = create_resource(
+        || (),
+        move |_| token_metadata_or_fallback(cans.clone(), user_principal, token_root),
+    );
+
+    view! {
+        <Suspense fallback=TokenClaimFallback>
+            {move || {
+                info.map(|info| {
+                    view! {
+                        <p>{format!("{} {}", info.balance.humanize(), info.symbol)}</p>
+                    }
+                })
+            }}
+        </Suspense>
+    }
+}
+
+#[component]
+pub fn ClaimTokensOrRedirectError(
+    token_root: Principal,
+    user_principal: Principal,
+) -> impl IntoView {
+    log::debug!("ClaimTokensOrRedirectError for {token_root}, {user_principal}");
     let auth_cans = authenticated_canisters();
+
+    let check_res = auth_cans.derive(
+        || (),
+        move |cans_wire, _| async move {
+            log::debug!("Checking if tokens need to be claimed for {token_root}");
+            let cans_wire = cans_wire?;
+            let res = claim_tokens_from_first_neuron_if_required(
+                cans_wire.clone(),
+                token_root,
+                user_principal,
+                true,
+            )
+            .await;
+            match res {
+                Ok(_) => Ok(()),  // No claim required, token balance can be displayed
+                Err(e) => Err(e), // Claim required or another error occurred
+            }
+        },
+    );
+
     let claim_res = auth_cans.derive(
```

Comment:
`claim_res` should be created only when required,
Resources execute as soon as you create them.